modularize the config module bootstrap

bit-aloo opened the pull request:
Currently, our `config` module is quite large over 3,000 lines, and handles a wide range of responsibilities. This PR aims to break it down into smaller, more focused submodules to improve readability and maintainability:

* **`toml`**: Introduces a dedicated `toml` submodule within the `config` module. Its sole purpose is to define configuration-related structs along with their corresponding deserialization logic. It also contains the `parse_inner` method, which serves as the central function for extracting relevant information from the TOML structs and constructing the final configuration.

* **`rust`, `dist`, `install`, `llvm`, `build`, `gcc`, and others**: Each of these modules contains TOML subsections specific to their domain, along with the logic necessary to convert them into parts of the final configuration struct.

* **`config/mod.rs`**: Contains shared types and enums used across multiple TOML subsections.

* **`config/config.rs`**: Houses the logic that integrates all the TOML subsections into the complete configuration struct.


r? @kobzol

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-19`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15128042336/job/42523649383) [(plain)](https://github.com/rust-lang/rust/commit/b8374d62fdb0c5125e395ba1a0ac7cec3497694a/checks/42523649383/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#19 exporting to docker image format
#19 sending tarball 20.8s done
#19 DONE 27.2s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-19]
[CI_JOB_NAME=x86_64-gnu-llvm-19]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-19', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--set', 'gcc.download-ci-gcc=true', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-19/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
[RUSTC-TIMING] proc_macro test:false 5.612
[RUSTC-TIMING] test test:false 9.605
    Finished `release` profile [optimized] target(s) in 1m 20s
##[endgroup]
[TIMING] core::build_steps::compile::Std { target: x86_64-unknown-linux-gnu, compiler: Compiler { stage: 1, host: x86_64-unknown-linux-gnu, forced_compiler: true }, crates: [], force_recompile: false, extra_rust_args: ["-Csymbol-mangling-version=v0", "-Cpanic=abort"], is_for_mir_opt_tests: false } -- 80.701
Testing GCC stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
 Downloading crates ...
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
[RUSTC-TIMING] boml test:false 0.783
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
[RUSTC-TIMING] y test:false 2.981
    Finished `release` profile [optimized] target(s) in 4.06s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-backend gcc --gcc-path /checkout/obj/build/x86_64-unknown-linux-gnu/ci-gcc/lib --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
`--gcc-path` was provided, ignoring config file. Using `/checkout/obj/build/x86_64-unknown-linux-gnu/ci-gcc/lib` as path for libgccjit
[BUILD] mini_core
[RUSTC-TIMING] mini_core test:false 0.206
[BUILD] example
[AOT] mini_core_hello_world
[RUSTC-TIMING] mini_core_hello_world test:false 0.184
---

error[E0432]: unresolved import `crate::core::config::TomlConfig`
  --> src/bootstrap/src/core/config/tests.rs:18:61
   |
18 | use crate::core::config::{LldMode, Target, TargetSelection, TomlConfig};
   |                                                             ^^^^^^^^^^ no `TomlConfig` in `core::config`
   |
   = help: consider importing this struct instead:
           crate::core::config::toml::TomlConfig

For more information about this error, try `rustc --explain E0432`.
```

</details>

Kobzol commented on src/bootstrap/src/core/config/toml.rs:
```diff
@@ -0,0 +1,544 @@
+#[serde(deny_unknown_fields, rename_all = "kebab-case")]
+pub(crate) struct TomlConfig {
+    #[serde(flatten)]
+    pub(crate) change_id: ChangeIdWrapper,
```
You can mark these with `pub(super)` to only make them available in the parent module, other parts of bootstrap probably shouldn't touch these anyway. This is likely true on more places than just here.

Kobzol reviewed (commented):
I like the general direction of this. Could you perhaps split the changes into multiple commits though? It would be easier to review. Thank you :)

onur-ozkan commented on src/bootstrap/src/core/config/toml.rs:
```diff
@@ -0,0 +1,544 @@
+    Id(usize),
+}
+
+/// Since we use `#[serde(deny_unknown_fields)]` on `TomlConfig`, we need a wrapper type
+/// for the "change-id" field to parse it even if other fields are invalid. This ensures
+/// that if deserialization fails due to other fields, we can still provide the changelogs
+/// to allow developers to potentially find the reason for the failure in the logs..
+#[derive(Deserialize, Default)]
+pub(crate) struct ChangeIdWrapper {
```
We have `types` module but this module also includes bunch of types. Right now it seems like there is no clear distinction between these modules and people will start putting things all around without having a clear clue. If we want to have module for per task, we have to make it super explicit/clear and have it well documented.

onur-ozkan commented on src/bootstrap/src/core/config/types.rs:
```diff
@@ -0,0 +1,524 @@
+    Full,
+}
+
+// NOTE: can't derive(Deserialize) because the intermediate trip through toml::Value only
+// deserializes i64, and derive() only generates visit_u64
+impl<'de> Deserialize<'de> for DebuginfoLevel {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        use serde::de::Error;
+
+        Ok(match Deserialize::deserialize(deserializer)? {
+            StringOrInt::String(s) if s == "none" => DebuginfoLevel::None,
+            StringOrInt::Int(0) => DebuginfoLevel::None,
+            StringOrInt::String(s) if s == "line-directives-only" => {
+                DebuginfoLevel::LineDirectivesOnly
+            }
+            StringOrInt::String(s) if s == "line-tables-only" => DebuginfoLevel::LineTablesOnly,
+            StringOrInt::String(s) if s == "limited" => DebuginfoLevel::Limited,
+            StringOrInt::Int(1) => DebuginfoLevel::Limited,
+            StringOrInt::String(s) if s == "full" => DebuginfoLevel::Full,
+            StringOrInt::Int(2) => DebuginfoLevel::Full,
+            StringOrInt::Int(n) => {
+                let other = serde::de::Unexpected::Signed(n);
+                return Err(D::Error::invalid_value(other, &"expected 0, 1, or 2"));
+            }
+            StringOrInt::String(s) => {
+                let other = serde::de::Unexpected::Str(&s);
+                return Err(D::Error::invalid_value(
+                    other,
+                    &"expected none, line-tables-only, limited, or full",
+                ));
+            }
+        })
+    }
+}
```
This is a parsing logic.

onur-ozkan commented on src/bootstrap/src/core/config/types.rs:
```diff
@@ -0,0 +1,524 @@
+use std::fmt::Display;
+use std::path::PathBuf;
+use std::str::FromStr;
+use std::{env, fmt};
+
+use serde::{Deserialize, Deserializer};
+use serde_derive::Deserialize;
+
+use crate::Path;
+use crate::utils::cache::{INTERNER, Interned};
+
```
Modules should have doc-comments to help https://github.com/rust-lang/rust/pull/141272#discussion_r2105780457.

onur-ozkan commented:
@rustbot author

jieyouxu commented:
cc @clubby789

jieyouxu commented:
Vibes-wise, organization seems fine to me.

onur-ozkan commented:
I still have [this](https://github.com/rust-lang/rust/pull/141272#discussion_r2105780457) concern. With the current approach, adding new things will always be unclear for some people. Even now, for example, we have `parsing.rs` and `types.rs` modules but some parsing logic has leaked into other modules and some types have leaked to some other modules as well.

What I would suggest instead is organizing the modules based on the configuration field. For example, `mod.rs` could represent the root level of the configuration and the rest could be split into the sub fields like  `rust.rs` for `[rust]`, `llvm.rs` for `[llvm], `target_specific.rs` for things like `[target.x86_64-unknown-linux-gnu]`, etc.

With this structure it would make it very explicit where things belong and anything that doesn't fit this rule can be easily caught during PR reviews.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15323843623/job/43113458909) [(plain)](https://github.com/rust-lang/rust/commit/2dc25e690663a6f0931e33a4fd7488e0660458c5/checks/43113458909/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

error: doc list item overindented
  --> src/bootstrap/src/core/config/toml/target.rs:10:5
   |
10 | //!     build target, which is is stored in the main [`Config`] structure.
   |     ^^^^ help: try using `  ` (2 spaces)
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#doc_overindented_list_items

error: doc list item overindented
  --> src/bootstrap/src/core/config/toml/target.rs:12:5
   |
12 | //!     applies it to the global [`Config`], ensuring proper path resolution, validation,
   |     ^^^^ help: try using `  ` (2 spaces)
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#doc_overindented_list_items

error: doc list item overindented
```

</details>

onur-ozkan commented on src/bootstrap/src/core/config/toml/macros.rs:
I think things like this can be defined in the root module. The same goes for other modules like `parsing.rs` and `common.rs`. The policy for creating modules under the toml directory should be very explicit and clear as I suggested above, which is having one module for per section in the bootstrap.toml file. So any module other than `mod.rs` should correspond to a section in the config file.

Kobzol replied:
This PR is already big, so let's go with the previous plan proposed by Onur, to have one file per config section. We can always split the root module later.

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-19-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15376611642/job/43262110080) [(plain)](https://github.com/rust-lang/rust/commit/1f52b291036a11b6f29a9361d386c9c54a605238/checks/43262110080/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling build_helper v0.1.0 (/checkout/src/build_helper)
error[E0599]: no function or associated item named `from_str` found for struct `PathBuf` in the current scope
##[error]   --> src/bootstrap/src/core/config/config.rs:745:46
    |
745 |         config.initial_sysroot = t!(PathBuf::from_str(
    |                                              ^^^^^^^^ function or associated item not found in `PathBuf`
    |
note: if you're trying to build a new `PathBuf` consider using one of the following associated functions:
      PathBuf::new
      PathBuf::with_capacity
---
17  + use std::str::FromStr;
    |
help: there is an associated function `from` with a similar name
    |
745 -         config.initial_sysroot = t!(PathBuf::from_str(
745 +         config.initial_sysroot = t!(PathBuf::from(
    |

For more information about this error, try `rustc --explain E0599`.
error: could not compile `bootstrap` (lib) due to 1 previous error
failed to run: /checkout/obj/build/aarch64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /checkout/src/bootstrap/Cargo.toml -Zroot-dir=/checkout --locked --features build-metrics
---
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
error[E0599]: no function or associated item named `from_str` found for struct `PathBuf` in the current scope
##[error]   --> src/bootstrap/src/core/config/config.rs:745:46
    |
745 |         config.initial_sysroot = t!(PathBuf::from_str(
    |                                              ^^^^^^^^ function or associated item not found in `PathBuf`
    |
note: if you're trying to build a new `PathBuf` consider using one of the following associated functions:
      PathBuf::new
      PathBuf::with_capacity
---
17  + use std::str::FromStr;
    |
help: there is an associated function `from` with a similar name
    |
745 -         config.initial_sysroot = t!(PathBuf::from_str(
745 +         config.initial_sysroot = t!(PathBuf::from(
    |

For more information about this error, try `rustc --explain E0599`.
error: could not compile `bootstrap` (lib) due to 1 previous error
failed to run: /checkout/obj/build/aarch64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /checkout/src/bootstrap/Cargo.toml -Zroot-dir=/checkout --locked --features build-metrics
---
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
error[E0599]: no function or associated item named `from_str` found for struct `PathBuf` in the current scope
##[error]   --> src/bootstrap/src/core/config/config.rs:745:46
    |
745 |         config.initial_sysroot = t!(PathBuf::from_str(
    |                                              ^^^^^^^^ function or associated item not found in `PathBuf`
    |
note: if you're trying to build a new `PathBuf` consider using one of the following associated functions:
      PathBuf::new
      PathBuf::with_capacity
---
17  + use std::str::FromStr;
    |
help: there is an associated function `from` with a similar name
    |
745 -         config.initial_sysroot = t!(PathBuf::from_str(
745 +         config.initial_sysroot = t!(PathBuf::from(
    |

For more information about this error, try `rustc --explain E0599`.
error: could not compile `bootstrap` (lib) due to 1 previous error
failed to run: /checkout/obj/build/aarch64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /checkout/src/bootstrap/Cargo.toml -Zroot-dir=/checkout --locked --features build-metrics
---
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
error[E0599]: no function or associated item named `from_str` found for struct `PathBuf` in the current scope
##[error]   --> src/bootstrap/src/core/config/config.rs:745:46
    |
745 |         config.initial_sysroot = t!(PathBuf::from_str(
    |                                              ^^^^^^^^ function or associated item not found in `PathBuf`
    |
note: if you're trying to build a new `PathBuf` consider using one of the following associated functions:
      PathBuf::new
      PathBuf::with_capacity
---
17  + use std::str::FromStr;
    |
help: there is an associated function `from` with a similar name
    |
745 -         config.initial_sysroot = t!(PathBuf::from_str(
745 +         config.initial_sysroot = t!(PathBuf::from(
    |

For more information about this error, try `rustc --explain E0599`.
error: could not compile `bootstrap` (lib) due to 1 previous error
failed to run: /checkout/obj/build/aarch64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /checkout/src/bootstrap/Cargo.toml -Zroot-dir=/checkout --locked --features build-metrics
---
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
error[E0599]: no function or associated item named `from_str` found for struct `PathBuf` in the current scope
##[error]   --> src/bootstrap/src/core/config/config.rs:745:46
    |
745 |         config.initial_sysroot = t!(PathBuf::from_str(
    |                                              ^^^^^^^^ function or associated item not found in `PathBuf`
    |
note: if you're trying to build a new `PathBuf` consider using one of the following associated functions:
      PathBuf::new
      PathBuf::with_capacity
---
17  + use std::str::FromStr;
    |
help: there is an associated function `from` with a similar name
    |
745 -         config.initial_sysroot = t!(PathBuf::from_str(
745 +         config.initial_sysroot = t!(PathBuf::from(
    |

For more information about this error, try `rustc --explain E0599`.
error: could not compile `bootstrap` (lib) due to 1 previous error
failed to run: /checkout/obj/build/aarch64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /checkout/src/bootstrap/Cargo.toml -Zroot-dir=/checkout --locked --features build-metrics
```

</details>

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-19-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15376839815/job/43262601621) [(plain)](https://github.com/rust-lang/rust/commit/1ec9ad36c3444f36ba26bd55b2dc4a96aeda8092/checks/43262601621/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling build_helper v0.1.0 (/checkout/src/build_helper)
error[E0599]: no function or associated item named `from_str` found for struct `PathBuf` in the current scope
##[error]   --> src/bootstrap/src/core/config/config.rs:745:46
    |
745 |         config.initial_sysroot = t!(PathBuf::from_str(
    |                                              ^^^^^^^^ function or associated item not found in `PathBuf`
    |
note: if you're trying to build a new `PathBuf` consider using one of the following associated functions:
      PathBuf::new
      PathBuf::with_capacity
---
17  + use std::str::FromStr;
    |
help: there is an associated function `from` with a similar name
    |
745 -         config.initial_sysroot = t!(PathBuf::from_str(
745 +         config.initial_sysroot = t!(PathBuf::from(
    |

For more information about this error, try `rustc --explain E0599`.
error: could not compile `bootstrap` (lib) due to 1 previous error
failed to run: /checkout/obj/build/aarch64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /checkout/src/bootstrap/Cargo.toml -Zroot-dir=/checkout --locked --features build-metrics
---
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
error[E0599]: no function or associated item named `from_str` found for struct `PathBuf` in the current scope
##[error]   --> src/bootstrap/src/core/config/config.rs:745:46
    |
745 |         config.initial_sysroot = t!(PathBuf::from_str(
    |                                              ^^^^^^^^ function or associated item not found in `PathBuf`
    |
note: if you're trying to build a new `PathBuf` consider using one of the following associated functions:
      PathBuf::new
      PathBuf::with_capacity
---
17  + use std::str::FromStr;
    |
help: there is an associated function `from` with a similar name
    |
745 -         config.initial_sysroot = t!(PathBuf::from_str(
745 +         config.initial_sysroot = t!(PathBuf::from(
    |

For more information about this error, try `rustc --explain E0599`.
error: could not compile `bootstrap` (lib) due to 1 previous error
failed to run: /checkout/obj/build/aarch64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /checkout/src/bootstrap/Cargo.toml -Zroot-dir=/checkout --locked --features build-metrics
---
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
error[E0599]: no function or associated item named `from_str` found for struct `PathBuf` in the current scope
##[error]   --> src/bootstrap/src/core/config/config.rs:745:46
    |
745 |         config.initial_sysroot = t!(PathBuf::from_str(
    |                                              ^^^^^^^^ function or associated item not found in `PathBuf`
    |
note: if you're trying to build a new `PathBuf` consider using one of the following associated functions:
      PathBuf::new
      PathBuf::with_capacity
---
17  + use std::str::FromStr;
    |
help: there is an associated function `from` with a similar name
    |
745 -         config.initial_sysroot = t!(PathBuf::from_str(
745 +         config.initial_sysroot = t!(PathBuf::from(
    |

For more information about this error, try `rustc --explain E0599`.
error: could not compile `bootstrap` (lib) due to 1 previous error
failed to run: /checkout/obj/build/aarch64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /checkout/src/bootstrap/Cargo.toml -Zroot-dir=/checkout --locked --features build-metrics
---
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
error[E0599]: no function or associated item named `from_str` found for struct `PathBuf` in the current scope
##[error]   --> src/bootstrap/src/core/config/config.rs:745:46
    |
745 |         config.initial_sysroot = t!(PathBuf::from_str(
    |                                              ^^^^^^^^ function or associated item not found in `PathBuf`
    |
note: if you're trying to build a new `PathBuf` consider using one of the following associated functions:
      PathBuf::new
      PathBuf::with_capacity
---
17  + use std::str::FromStr;
    |
help: there is an associated function `from` with a similar name
    |
745 -         config.initial_sysroot = t!(PathBuf::from_str(
745 +         config.initial_sysroot = t!(PathBuf::from(
    |

For more information about this error, try `rustc --explain E0599`.
error: could not compile `bootstrap` (lib) due to 1 previous error
failed to run: /checkout/obj/build/aarch64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /checkout/src/bootstrap/Cargo.toml -Zroot-dir=/checkout --locked --features build-metrics
---
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
error[E0599]: no function or associated item named `from_str` found for struct `PathBuf` in the current scope
##[error]   --> src/bootstrap/src/core/config/config.rs:745:46
    |
745 |         config.initial_sysroot = t!(PathBuf::from_str(
    |                                              ^^^^^^^^ function or associated item not found in `PathBuf`
    |
note: if you're trying to build a new `PathBuf` consider using one of the following associated functions:
      PathBuf::new
      PathBuf::with_capacity
---
17  + use std::str::FromStr;
    |
help: there is an associated function `from` with a similar name
    |
745 -         config.initial_sysroot = t!(PathBuf::from_str(
745 +         config.initial_sysroot = t!(PathBuf::from(
    |

For more information about this error, try `rustc --explain E0599`.
error: could not compile `bootstrap` (lib) due to 1 previous error
failed to run: /checkout/obj/build/aarch64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /checkout/src/bootstrap/Cargo.toml -Zroot-dir=/checkout --locked --features build-metrics
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #119899) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bit-aloo committed:
This commit adds a `toml` module that represents various subsections of `bootstrap.toml`.
The module primarily defines the TOML representations of these subsections and their corresponding
configuration parsing implementations, which are ultimately integrated into the main configuration.

- `mod.rs` – Serves as the entry point for the TOML configuration and defines the `TomlConfig` struct along with its parsing implementation.
- `rust.rs` – Defines the `Rust` subsection struct and its configuration parsing implementation.
- `target.rs` – Defines the `Target` subsection struct and its configuration parsing implementation.
- `llvm.rs` – Defines the `Llvm` subsection struct and its configuration parsing implementation.
- `install.rs` – Defines the `Install` subsection struct and its configuration parsing implementation.
- `gcc.rs` – Defines the `Gcc` subsection struct and its configuration parsing implementation.
- `dist.rs` – Defines the `Dist` subsection struct and its configuration parsing implementation.
- `build.rs` – Defines the `Build` subsection struct and its configuration parsing implementation.

bit-aloo committed:
move target selection to its separate module

bit-aloo committed:
Move all common types, macros and utility method to config/mod.rs

bit-aloo committed:
add config.rs with new module structuring

bit-aloo committed:
correct the imports in flags, tests and download

onur-ozkan reviewed (approved):
LGTM. Thank you so much for looking into this!

Kobzol commented:
No, I'm good. Thank you, great work!

@bors r+

bors commented:
:pushpin: Commit 6ec7ec96892d45792242e95164a7eb903bcd3f76 has been approved by `Kobzol`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Kobzol 6ec7ec96892d45792242e95164a7eb903bcd3f76 -->
<!-- homu: {"type":"Approved","sha":"6ec7ec96892d45792242e95164a7eb903bcd3f76","approver":"Kobzol","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

matthiaskrgr commented:
@bors r-
https://github.com/rust-lang/rust/pull/142059#issuecomment-2943610220

Kobzol commented:
Hmm, we should probably run `x doc bootstrap` in PR CI, this is not the first time this has bitten us, and we will hopefully document more and more things in bootstrap now.

@bors r+ rollup=iffy

bors commented:
:pushpin: Commit 3667dbd5f9480bcf114a06d1093385a77211085d has been approved by `Kobzol`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Kobzol 3667dbd5f9480bcf114a06d1093385a77211085d -->
<!-- homu: {"type":"Approved","sha":"3667dbd5f9480bcf114a06d1093385a77211085d","approver":"Kobzol","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
